File: frontend/lib/api/user-settings.ts
/**
 * User Settings API Client
 * 
 * Provides methods for managing user preferences:
 * - Notifications
 * - Security
 * - Appearance
 * - LLM configuration
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

// ============================================================================
// Types
// ============================================================================

export type EmailDigestFrequency = 'realtime' | 'hourly' | 'daily' | 'weekly' | 'never';
export type TwoFactorMethod = 'app' | 'sms' | 'email';
export type ThemeMode = 'light' | 'dark' | 'system';

export interface NotificationSettings {
  email_enabled: boolean;
  email_address?: string;
  email_on_workflow_complete: boolean;
  email_on_workflow_error: boolean;
  email_on_agent_alert: boolean;
  email_digest_frequency: EmailDigestFrequency;
  
  in_app_enabled: boolean;
  in_app_on_workflow_complete: boolean;
  in_app_on_workflow_error: boolean;
  in_app_on_agent_alert: boolean;
  in_app_on_system_update: boolean;
  
  slack_enabled: boolean;
  slack_webhook_url?: string;
  slack_on_workflow_complete: boolean;
  slack_on_workflow_error: boolean;
  
  browser_enabled: boolean;
  browser_on_workflow_complete: boolean;
  browser_on_workflow_error: boolean;
}

export interface SecuritySettings {
  require_strong_password: boolean;
  password_min_length: number;
  
  session_timeout: number;
  single_session: boolean;
  
  two_factor_enabled: boolean;
  two_factor_method: TwoFactorMethod;
  
  api_rate_limit_enabled: boolean;
  api_rate_limit_requests: number;
  api_rate_limit_window: number;
  
  audit_logging_enabled: boolean;
  log_api_calls: boolean;
  log_workflow_executions: boolean;
  log_login_attempts: boolean;
}

export interface AppearanceSettings {
  theme: ThemeMode;
  accent_color: string;
  
  font_size: number;
  font_family: string;
  
  sidebar_collapsed: boolean;
  compact_mode: boolean;
  show_breadcrumbs: boolean;
  
  editor_font_size: number;
  editor_line_numbers: boolean;
  editor_word_wrap: boolean;
  editor_minimap: boolean;
  
  reduce_motion: boolean;
  high_contrast: boolean;
}

export interface LLMSettings {
  default_provider: string;
  default_model: string;
  
  ollama_enabled: boolean;
  ollama_base_url: string;
  ollama_default_model: string;
}

export interface AllUserSettings {
  notifications: NotificationSettings;
  security: SecuritySettings;
  appearance: AppearanceSettings;
  llm: LLMSettings;
  updated_at?: string;
}

// ============================================================================
// API Client
// ============================================================================

class UserSettingsAPI {
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const token = localStorage.getItem('token');
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
      ...options.headers,
    };

    const response = await fetch(`${API_BASE_URL}${endpoint}`, {
      ...options,
      headers,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: 'Request failed' }));
      throw new Error(error.detail || `HTTP ${response.status}`);
    }

    return response.json();
  }

  // Get all settings
  async getAllSettings(): Promise<AllUserSettings> {
    return this.request('/api/agent-builder/user-settings');
  }

  // Update all settings
  async updateAllSettings(settings: AllUserSettings): Promise<AllUserSettings> {
    return this.request('/api/agent-builder/user-settings', {
      method: 'PUT',
      body: JSON.stringify(settings),
    });
  }

  // Notification settings
  async getNotificationSettings(): Promise<NotificationSettings> {
    return this.request('/api/agent-builder/user-settings/notifications');
  }

  async updateNotificationSettings(settings: NotificationSettings): Promise<NotificationSettings> {
    return this.request('/api/agent-builder/user-settings/notifications', {
      method: 'PUT',
      body: JSON.stringify(settings),
    });
  }

  // Security settings
  async getSecuritySettings(): Promise<SecuritySettings> {
    return this.request('/api/agent-builder/user-settings/security');
  }

  async updateSecuritySettings(settings: SecuritySettings): Promise<SecuritySettings> {
    return this.request('/api/agent-builder/user-settings/security', {
      method: 'PUT',
      body: JSON.stringify(settings),
    });
  }

  // Appearance settings
  async getAppearanceSettings(): Promise<AppearanceSettings> {
    return this.request('/api/agent-builder/user-settings/appearance');
  }

  async updateAppearanceSettings(settings: AppearanceSettings): Promise<AppearanceSettings> {
    return this.request('/api/agent-builder/user-settings/appearance', {
      method: 'PUT',
      body: JSON.stringify(settings),
    });
  }

  // LLM settings
  async getLLMSettings(): Promise<LLMSettings> {
    return this.request('/api/agent-builder/user-settings/llm');
  }

  async updateLLMSettings(settings: LLMSettings): Promise<LLMSettings> {
    return this.request('/api/agent-builder/user-settings/llm', {
      method: 'PUT',
      body: JSON.stringify(settings),
    });
  }

  // Reset settings
  async resetSettings(category?: string): Promise<{ message: string }> {
    const params = category ? `?category=${category}` : '';
    return this.request(`/api/agent-builder/user-settings/reset${params}`, {
      method: 'POST',
    });
  }

  // Export settings
  async exportSettings(): Promise<{ user_id: string; exported_at: string; settings: any }> {
    return this.request('/api/agent-builder/user-settings/export', {
      method: 'POST',
    });
  }

  // Import settings
  async importSettings(settings: Record<string, any>): Promise<{ message: string }> {
    return this.request('/api/agent-builder/user-settings/import', {
      method: 'POST',
      body: JSON.stringify(settings),
    });
  }
}

export const userSettingsAPI = new UserSettingsAPI();

// ============================================================================
// Default Settings
// ============================================================================

export const defaultNotificationSettings: NotificationSettings = {
  email_enabled: true,
  email_address: '',
  email_on_workflow_complete: false,
  email_on_workflow_error: true,
  email_on_agent_alert: true,
  email_digest_frequency: 'daily',
  
  in_app_enabled: true,
  in_app_on_workflow_complete: true,
  in_app_on_workflow_error: true,
  in_app_on_agent_alert: true,
  in_app_on_system_update: true,
  
  slack_enabled: false,
  slack_webhook_url: '',
  slack_on_workflow_complete: false,
  slack_on_workflow_error: true,
  
  browser_enabled: false,
  browser_on_workflow_complete: false,
  browser_on_workflow_error: true,
};

export const defaultSecuritySettings: SecuritySettings = {
  require_strong_password: true,
  password_min_length: 8,
  
  session_timeout: 30,
  single_session: false,
  
  two_factor_enabled: false,
  two_factor_method: 'app',
  
  api_rate_limit_enabled: true,
  api_rate_limit_requests: 100,
  api_rate_limit_window: 60,
  
  audit_logging_enabled: true,
  log_api_calls: true,
  log_workflow_executions: true,
  log_login_attempts: true,
};

export const defaultAppearanceSettings: AppearanceSettings = {
  theme: 'system',
  accent_color: '#3b82f6',
  
  font_size: 14,
  font_family: 'system',
  
  sidebar_collapsed: false,
  compact_mode: false,
  show_breadcrumbs: true,
  
  editor_font_size: 14,
  editor_line_numbers: true,
  editor_word_wrap: true,
  editor_minimap: true,
  
  reduce_motion: false,
  high_contrast: false,
};

export const defaultLLMSettings: LLMSettings = {
  default_provider: 'openai',
  default_model: 'gpt-4o-mini',
  
  ollama_enabled: false,
  ollama_base_url: 'http://localhost:11434',
  ollama_default_model: 'llama3.1',
};

File: backend/api/agent_builder/user_settings.py
"""
User Settings API for Agent Builder

Provides endpoints for managing user preferences including:
- Notification settings
- Security settings
- Appearance settings
- LLM configuration
"""

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime
from enum import Enum
import logging
from sqlalchemy.orm import Session

from backend.db.database import get_db
from backend.core.auth_dependencies import get_current_user
from backend.db.models.user import User

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/agent-builder/user-settings", tags=["User Settings"])


# ============================================================================
# Models
# ============================================================================

class EmailDigestFrequency(str, Enum):
    REALTIME = "realtime"
    HOURLY = "hourly"
    DAILY = "daily"
    WEEKLY = "weekly"
    NEVER = "never"


class TwoFactorMethod(str, Enum):
    APP = "app"
    SMS = "sms"
    EMAIL = "email"


class ThemeMode(str, Enum):
    LIGHT = "light"
    DARK = "dark"
    SYSTEM = "system"


class NotificationSettings(BaseModel):
    """Notification preferences"""
    email_enabled: bool = True
    email_address: Optional[str] = None
    email_on_workflow_complete: bool = False
    email_on_workflow_error: bool = True
    email_on_agent_alert: bool = True
    email_digest_frequency: EmailDigestFrequency = EmailDigestFrequency.DAILY
    
    in_app_enabled: bool = True
    in_app_on_workflow_complete: bool = True
    in_app_on_workflow_error: bool = True
    in_app_on_agent_alert: bool = True
    in_app_on_system_update: bool = True
    
    slack_enabled: bool = False
    slack_webhook_url: Optional[str] = None
    slack_on_workflow_complete: bool = False
    slack_on_workflow_error: bool = True
    
    browser_enabled: bool = False
    browser_on_workflow_complete: bool = False
    browser_on_workflow_error: bool = True


class SecuritySettings(BaseModel):
    """Security preferences"""
    require_strong_password: bool = True
    password_min_length: int = Field(default=8, ge=6, le=32)
    
    session_timeout: int = Field(default=30, ge=5, le=480)  # minutes
    single_session: bool = False
    
    two_factor_enabled: bool = False
    two_factor_method: TwoFactorMethod = TwoFactorMethod.APP
    
    api_rate_limit_enabled: bool = True
    api_rate_limit_requests: int = Field(default=100, ge=10, le=10000)
    api_rate_limit_window: int = Field(default=60, ge=10, le=3600)  # seconds
    
    audit_logging_enabled: bool = True
    log_api_calls: bool = True
    log_workflow_executions: bool = True
    log_login_attempts: bool = True


class AppearanceSettings(BaseModel):
    """Appearance preferences"""
    theme: ThemeMode = ThemeMode.SYSTEM
    accent_color: str = Field(default="#3b82f6", pattern=r"^#[0-9a-fA-F]{6}$")
    
    font_size: int = Field(default=14, ge=10, le=24)
    font_family: str = "system"
    
    sidebar_collapsed: bool = False
    compact_mode: bool = False
    show_breadcrumbs: bool = True
    
    editor_font_size: int = Field(default=14, ge=10, le=24)
    editor_line_numbers: bool = True
    editor_word_wrap: bool = True
    editor_minimap: bool = True
    
    reduce_motion: bool = False
    high_contrast: bool = False


class LLMSettings(BaseModel):
    """LLM configuration"""
    default_provider: str = "openai"
    default_model: str = "gpt-4o-mini"
    
    ollama_enabled: bool = False
    ollama_base_url: str = "http://localhost:11434"
    ollama_default_model: str = "llama3.1"
    
    # API keys are stored separately for security


class AllUserSettings(BaseModel):
    """Combined user settings"""
    notifications: NotificationSettings = NotificationSettings()
    security: SecuritySettings = SecuritySettings()
    appearance: AppearanceSettings = AppearanceSettings()
    llm: LLMSettings = LLMSettings()
    updated_at: Optional[datetime] = None


# In-memory storage (in production, use database)
_user_settings: Dict[int, Dict[str, Any]] = {}


# ============================================================================
# Endpoints
# ============================================================================

@router.get("", response_model=AllUserSettings)
async def get_all_settings(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get all user settings."""
    user_id = current_user.id
    settings = _user_settings.get(user_id, {})
    
    return AllUserSettings(
        notifications=NotificationSettings(**settings.get("notifications", {})),
        security=SecuritySettings(**settings.get("security", {})),
        appearance=AppearanceSettings(**settings.get("appearance", {})),
        llm=LLMSettings(**settings.get("llm", {})),
        updated_at=settings.get("updated_at")
    )


@router.put("", response_model=AllUserSettings)
async def update_all_settings(
    settings_data: AllUserSettings,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update all user settings."""
    user_id = current_user.id
    
    _user_settings[user_id] = {
        "notifications": settings_data.notifications.model_dump(),
        "security": settings_data.security.model_dump(),
        "appearance": settings_data.appearance.model_dump(),
        "llm": settings_data.llm.model_dump(),
        "updated_at": datetime.utcnow()
    }
    
    logger.info(f"User {user_id} updated all settings")
    
    return AllUserSettings(
        **_user_settings[user_id],
        updated_at=_user_settings[user_id]["updated_at"]
    )


@router.get("/notifications", response_model=NotificationSettings)
async def get_notification_settings(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get notification settings."""
    user_id = current_user.id
    settings = _user_settings.get(user_id, {}).get("notifications", {})
    return NotificationSettings(**settings)


@router.put("/notifications", response_model=NotificationSettings)
async def update_notification_settings(
    settings_data: NotificationSettings,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update notification settings."""
    user_id = current_user.id
    
    if user_id not in _user_settings:
        _user_settings[user_id] = {}
    
    _user_settings[user_id]["notifications"] = settings_data.model_dump()
    _user_settings[user_id]["updated_at"] = datetime.utcnow()
    
    logger.info(f"User {user_id} updated notification settings")
    return settings_data


@router.get("/security", response_model=SecuritySettings)
async def get_security_settings(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get security settings."""
    user_id = current_user.id
    settings = _user_settings.get(user_id, {}).get("security", {})
    return SecuritySettings(**settings)


@router.put("/security", response_model=SecuritySettings)
async def update_security_settings(
    settings_data: SecuritySettings,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update security settings."""
    user_id = current_user.id
    
    if user_id not in _user_settings:
        _user_settings[user_id] = {}
    
    _user_settings[user_id]["security"] = settings_data.model_dump()
    _user_settings[user_id]["updated_at"] = datetime.utcnow()
    
    logger.info(f"User {user_id} updated security settings")
    return settings_data


@router.get("/appearance", response_model=AppearanceSettings)
async def get_appearance_settings(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get appearance settings."""
    user_id = current_user.id
    settings = _user_settings.get(user_id, {}).get("appearance", {})
    return AppearanceSettings(**settings)


@router.put("/appearance", response_model=AppearanceSettings)
async def update_appearance_settings(
    settings_data: AppearanceSettings,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update appearance settings."""
    user_id = current_user.id
    
    if user_id not in _user_settings:
        _user_settings[user_id] = {}
    
    _user_settings[user_id]["appearance"] = settings_data.model_dump()
    _user_settings[user_id]["updated_at"] = datetime.utcnow()
    
    logger.info(f"User {user_id} updated appearance settings")
    return settings_data


@router.get("/llm", response_model=LLMSettings)
async def get_llm_settings(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get LLM configuration settings."""
    user_id = current_user.id
    settings = _user_settings.get(user_id, {}).get("llm", {})
    return LLMSettings(**settings)


@router.put("/llm", response_model=LLMSettings)
async def update_llm_settings(
    settings_data: LLMSettings,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update LLM configuration settings."""
    user_id = current_user.id
    
    if user_id not in _user_settings:
        _user_settings[user_id] = {}
    
    _user_settings[user_id]["llm"] = settings_data.model_dump()
    _user_settings[user_id]["updated_at"] = datetime.utcnow()
    
    logger.info(f"User {user_id} updated LLM settings")
    return settings_data


@router.post("/reset")
async def reset_settings(
    category: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Reset settings to defaults.
    
    Args:
        category: Optional category to reset (notifications, security, appearance, llm).
                  If not provided, resets all settings.
    """
    user_id = current_user.id
    
    if category:
        if category not in ["notifications", "security", "appearance", "llm"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid category: {category}"
            )
        
        if user_id in _user_settings and category in _user_settings[user_id]:
            del _user_settings[user_id][category]
            logger.info(f"User {user_id} reset {category} settings")
    else:
        if user_id in _user_settings:
            del _user_settings[user_id]
            logger.info(f"User {user_id} reset all settings")
    
    return {"message": f"Settings reset successfully", "category": category or "all"}


@router.post("/export")
async def export_settings(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Export all user settings as JSON."""
    user_id = current_user.id
    settings = _user_settings.get(user_id, {})
    
    return {
        "user_id": str(user_id),
        "exported_at": datetime.utcnow().isoformat(),
        "settings": settings
    }


@router.post("/import")
async def import_settings(
    settings_data: Dict[str, Any],
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Import user settings from JSON."""
    user_id = current_user.id
    
    try:
        # Validate each category
        if "notifications" in settings_data:
            NotificationSettings(**settings_data["notifications"])
        if "security" in settings_data:
            SecuritySettings(**settings_data["security"])
        if "appearance" in settings_data:
            AppearanceSettings(**settings_data["appearance"])
        if "llm" in settings_data:
            LLMSettings(**settings_data["llm"])
        
        # Import settings
        _user_settings[user_id] = {
            **settings_data,
            "updated_at": datetime.utcnow()
        }
        
        logger.info(f"User {user_id} imported settings")
        
        return {"message": "Settings imported successfully"}
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid settings format: {str(e)}"
        )
